### PR TITLE
API

### LINKED ISSUE #4: validate parameters before passing to backend

### LINKED ISSUE #6: implement for API

### PR DESCRIPTION
**Pull request by dsavchenko**
Address #6 and #4

### CODE REVIEW

File: dispatcher_plugin_gw/dataserver_dispatcher.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import time
+from xml.etree.ElementTree import Comment
```
Comment (volodymyrss) [inline dispatcher_plugin_gw/dataserver_dispatcher.py]: what's this, auto-import?

File: dispatcher_plugin_gw/products.py [hunk 2]
```diff
@@ -193,22 +233,37 @@ def write(self):
...
+        
     
     def get_plot(self):
         script = ''
         div = f'<img src="data:image/svg+xml;base64,{self.imagedata}">'
         return script, div
     
-    def get_catalog_dict(self):
+    def get_catalog_dict(self, api=False):
+        #TODO: add metadata (tamerange etc)
```
Comment (volodymyrss) [inline dispatcher_plugin_gw/products.py]: timerange

**Review by volodymyrss — CHANGES_REQUESTED** [review]
Maybe could you add a test with dispatcher fixture? See in other plugins.

**Comment by volodymyrss** [comment]
But as you wish really. Thanks!

File: dispatcher_plugin_gw/dataserver_dispatcher.py (hunk 1 above)
Comment (dsavchenko) [reply dispatcher_plugin_gw/dataserver_dispatcher.py]: I think so. Clearly, not needed, thanks.

**Comment by dsavchenko** [comment]
> Maybe could you add a test with dispatcher fixture? See in other plugins.

I am trying to grasp, how tests are implemented for the integral plugin. As I see, no backend is started (for the Antares I run backend container in Github actions) and there are comments like _something already done at backend_ here and there, but I don't see, where these mock results are fed to the dispatcher fixture. 
Could you explain a bit the approach?

**Comment by volodymyrss** [comment]
> 

We do not have a mock backend for integral. Instead we have dummy products. Since the backend is too heavy.

In this case, something just like you did for antares would be sufficient, no?

https://github.com/oda-hub/dispatcher-plugin-antares/blob/master/tests/test_plugin.py

We should take care to add it to there as well:

https://github.com/oda-hub/dispatcher-app/blob/master/.github/workflows/python-plugins.yml

File: .github/python-app.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    - name: Lint with flake8
+      run: |
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics || echo "failed lint"
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics || echo "failed second lint"
+
+    - name: PyLint - more lint
+      run: |
+        pylint -E dispatcher_plugin_antares || echo 'this is fine!'
```
Comment (volodymyrss) [inline .github/python-app.yml]:
```suggestion
        pylint -E dispatcher_plugin_gw || echo 'this is fine!'
```

File: .github/python-app.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      run: |
+        mypy dispatcher_plugin_gw --ignore-missing-imports || echo 'this is fine too!'
+
+    - name: Test with pytest
+      run: |
+        python -m coverage run --source=dispatcher_plugin_gw -m pytest tests -sv --full-trace --log-cli-level=DEBUG
+
+    # I wonder how this interacts with the Github App. But this is not only for PR. Also this does not always work
+    - name: Codecov
+      uses: codecov/codecov-action@v1.3.1
```
Comment (volodymyrss) [inline .github/python-app.yml]: seems like this version deprecates tomorrow (https://github.com/marketplace/actions/codecov), we need to migrate

File: .github/python-app.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
```
Comment (volodymyrss) [inline .github/python-app.yml]: the file should be in `.github/workflows/python-app.yml`, else you can see it does not actually run.

File: dispatcher_plugin_gw/dataserver_dispatcher.py [hunk 6]
```diff
@@ -111,25 +110,34 @@ def run_query(self,
...
         else:
-            query_out.set_failed('Error in the backend', 
+            try:
+                query_out.set_failed('Error in the backend', 
                                  message='connection status code: ' + str(res.status_code), 
                                  extra_message=res.json()['exceptions'][0])
+            except:
+                query_out.set_failed('Error in the backend', 
+                                 message='connection status code: ' + str(res.status_code), 
+                                )
```
Comment (burnout87) [inline dispatcher_plugin_gw/dataserver_dispatcher.py]: perhaps as `extra_message` the `res.text` could be used.

File: dispatcher_plugin_gw/queries.py [hunk 7]
```diff
@@ -7,6 +7,14 @@
...
 
 from .products import SkymapProduct, SpectrogramProduct, StrainProduct
+from astropy.time import Time as astropyTime
+
+def check_time_interval(T1, T2, maxinterval=60):
+    t1 = astropyTime(T1, format='isot')
+    t2 = astropyTime(T2, format='isot')
+    delta = t2 - t1
+    if delta.sec > maxinterval:
+        raise ValueError(f'Too long time interval. Current limit is {maxinterval}s')
```
Comment (burnout87) [inline dispatcher_plugin_gw/queries.py:17]: can the `astropyTimeDelta` or the `TimeDelta` parameter be used here?

**Review by burnout87 — COMMENTED** [review]
just a couple of suggestions

File: dispatcher_plugin_gw/queries.py (hunk 7 above)
Comment (dsavchenko) [reply dispatcher_plugin_gw/queries.py:17]: delta here is actually astropy.time.TimeDelta

File: .github/python-app.yml (hunk 4 above)
Comment (dsavchenko) [reply .github/python-app.yml]: I actually just copied this from other repo. Will try to change to v2. 
There is also a token specified here, but isn't added to secrets (but do we need it for the public repo?). And no base report.
Could you manage it, please?

**Review by volodymyrss — APPROVED** [review]

**Pull request merged**